Docs: ✨ Add comprehensive GitHub Copilot instructions with verified sandbox workflow and enhanced build commands (#32339)
storybookjs/storybook: .github/copilot-instructions.md

- **Development servers**: Use async mode or timeout commands

### Example Bash Commands
```bash
# Safe compilation with proper timeout
bash(command="cd /path/to/storybook && yarn task --task compile", timeout=300, async=false)

# Start development server with timeout to prevent hanging
bash(command="cd /path/to/storybook/code && timeout 30s yarn storybook:ui", timeout=45, async=false)

# Use async for interactive or long-running commands
bash(command="cd /path/to/storybook/code && yarn storybook:ui", async=true)
```

## Sandbox Environments

### Available Templates
The `code/sandbox/` directory contains pre-built sandbox environments:
- `react-vite-default-ts` - React with Vite and TypeScript
- `angular-cli-default-ts` - Angular CLI with TypeScript
- `svelte-vite-default-ts` - Svelte with Vite and TypeScript
- `vue3-vite-default-ts` - Vue 3 with Vite and TypeScript
- And many more...

### Using Existing Sandboxes
Since sandbox creation fails due to API limits, use existing sandbox directories for testing:
```bash
# Navigate to an existing sandbox
cd code/sandbox/react-vite-default-ts
# Install dependencies if needed
yarn install
# Start the sandbox
yarn storybook

> ```

Reviewer: @copilot The section on sandbox environments is wrong. The `code/sandbox` directory is almost completely irrelevant. See my other comment:

> you have to first generate a sandbox with the framework/builder+renderer combination that you need to test out your stuff using yarn task --task sandbox and selecting the appropriate sandbox. Eg. if you're making changes to Svelte code, you probably want to select the svelte-vite/defaults-ts sandbox. Then you can cd sandbox/svelte-vite-default-ts and start Storybook from there or whatever.
> 
> Probably try this out to see how it works before you document it here.

# GitHub Copilot Instructions for Storybook

This document provides comprehensive instructions for GitHub Copilot when working on the Storybook repository.

## Repository Overview

Storybook is a large monorepo built with TypeScript, React, and various other frameworks. The main codebase is located in the `code/` directory, with additional tooling in `scripts/`.

## System Requirements

- **Node.js**: 22.16.0 (see `.nvmrc`)
- **Package Manager**: Yarn 4.9.1
- **Operating System**: Linux/macOS (CI environment)

## Repository Structure

```
storybook/
├── .github/           # GitHub configurations and workflows
├── code/              # Main monorepo codebase

> │   ├── addons/        # Storybook addons
> │   ├── frameworks/    # Framework integrations
> │   ├── lib/           # Core libraries
> │   ├── builders/      # Build tools
> │   └── sandbox/       # Internal build artifacts (not for testing)

Reviewer: ```suggestion
│   ├── .storybook/    # Configuration for internal UI Storybook
│   ├── core/          # Core Storybook package
│   ├── lib/           # Core supporting libraries
│   ├── addons/        # Core Storybook addons
│   ├── builders/      # Builder integrations
│   ├── renderers/     # Renderer integrations
│   ├── frameworks/    # Framework integrations
│   ├── presets/       # Preset packages for Webpack-based integrations
│   └── sandbox/       # Internal build artifacts (not useful for anything, ignore)
```

├── sandbox/           # Generated sandbox environments (created by yarn task --task sandbox)
├── scripts/           # Build and development scripts
├── docs/              # Documentation
└── test-storybooks/   # Test configurations
```

## Essential Commands and Build Times

### Installation & Setup
```bash
# Install all dependencies (run from repository root)
yarn i
# Time: ~2.5 minutes
# Timeout: Use 300+ seconds for bash commands
```

### Compilation
```bash
# Compile all packages (run from repository root)
yarn task --task compile
# Time: ~3 minutes (tested: 3m0.729s)
# Timeout: Use 300+ seconds for bash commands
```

### Linting
```bash
# Run all linting checks (run from repository root)
yarn lint
# Time: ~4 minutes
# Timeout: Use 300+ seconds for bash commands
```

> 

Reviewer: @copilot add a section on Type checking too, using `yarn task --task check`

### Development Server
```bash
# Start Storybook UI development server (run from code/ directory)
cd code && yarn storybook:ui
# Time: ~2.26 seconds startup time
# Serves on: http://localhost:6006/

> # Note: This runs indefinitely - use timeout or async mode

Reviewer: ```suggestion
# Note: This runs indefinitely - use timeout or async mode
# Note: This requires the repository to be compiled first, see Compilation above
```

# Build Storybook UI for production (run from code/ directory)
cd code && yarn storybook:ui:build
# Time: ~1m 46s
# Output: code/storybook-static/

> # Note: This does NOT run indefinitely

Reviewer: ```suggestion
# Note: This does NOT run indefinitely
# Note: This requires the repository to be compiled first, see Compilation above
```

```

### Testing
```bash
# Run all tests (run from code/ directory)
cd code && yarn test
# Time: Variable, depends on test scope
# Watch mode for continuous testing
cd code && yarn test:watch

# Storybook UI specific tests
cd code && yarn storybook:vitest

# Available task-based testing commands
yarn task --task e2e-tests-build    # E2E tests for built Storybook
yarn task --task e2e-tests-dev      # E2E tests for dev server
yarn task --task test-runner-build  # Test runner for built Storybook
yarn task --task test-runner-dev    # Test runner for dev server
yarn task --task smoke-test         # Basic smoke tests
yarn task --task vitest-test        # Vitest integration tests
```

## Important Warnings and Limitations

### Commands to Avoid
- **DO NOT RUN**: `yarn task --task dev` - This starts a permanent development server that runs indefinitely and will cause timeouts
- **DO NOT RUN**: `yarn start` - This also starts a long-running development server

> - **USE WITH CAUTION**: Any `*-dev` task commands (e2e-tests-dev, test-runner-dev) - These may start long-running servers

Reviewer: ```suggestion
```

### Available Task Commands
The repository includes 20 task scripts in `scripts/tasks/`:
- `bench` - Performance benchmarking
- `build` - Package building
- `check` - Package validation
- `chromatic` - Visual testing with Chromatic
- `compile` - TypeScript compilation
- `dev` - Development server (AVOID - runs indefinitely)
- `e2e-tests-build` - E2E tests for built Storybook
- `e2e-tests-dev` - E2E tests for dev server
- `generate` - Code generation
- `install` - Dependency installation
- `publish` - Package publishing
- `run-registry` - Local npm registry
- `sandbox` - Sandbox creation (fails due to API limits)
- `serve` - Static serving
- `smoke-test` - Basic functionality tests
- `sync-docs` - Documentation synchronization
- `test-runner-build` - Test runner for built Storybook
- `test-runner-dev` - Test runner for dev server
- `vitest-test` - Vitest integration tests

### Known Issues
1. **GitHub API Rate Limiting**: Sandbox creation fails with 403 Forbidden errors due to GitHub API limits in CI environments
   ```bash
   # This will fail in CI environments:
   yarn task --task sandbox --template react-vite/default-ts

>    ```

Reviewer: @copilot try this again, I've added github.com to the agent allowlist. If it works, remember to update all the references to this API limit throughout this doc.

2. **Dependency Warnings**: The build process shows many peer dependency warnings, but these are expected and don't prevent successful builds

3. **Large Build Times**: Most build operations take several minutes - always use appropriate timeouts

4. **Sandbox Generation**: Currently unreliable in CI environments - use Storybook UI for testing changes instead

## Recommended Development Workflow

### For Code Changes
1. Install dependencies: `yarn i` (if needed)
2. Compile packages: `yarn task --task compile`

> 3. Run linting: `yarn lint`
> 4. Make your changes
> 5. Test changes with: `cd code && yarn storybook:ui`
> 6. Run relevant tests: `cd code && yarn test`

Reviewer: ```suggestion
3. Make your changes
4. Compile packages with `cd code && yarn task --task compile`
5. Test changes with: `cd code && yarn storybook:ui:build`
6. Run relevant tests: `cd code && yarn test`
```

### For Testing UI Changes
1. Generate a sandbox with: `yarn task --task sandbox --template [framework-template]` (may fail due to API limits)
2. If sandbox generation fails, use Storybook UI: `cd code && yarn storybook:ui`
3. Access at http://localhost:6006/

> ### For Addon/Framework Development
> 1. Navigate to the relevant package in `code/addons/` or `code/frameworks/`
> 2. Make changes to source files
> 3. Rebuild with compilation command
> 4. Test in Storybook UI

Reviewer: ```suggestion
### For Addon/Framework/Renderers Development
1. Navigate to the relevant package in `code/addons/`, `code/frameworks/` or `code/renderers/`
2. Make changes to source files
3. Rebuild with compilation command
4. Generate a sandbox that matches the framework/renderer being worked on
5. Test with the appropriate test tasks
```
